### DISCUSSION TITLE
Proposal: A way to eliminate CS0031

### DISCUSSION THREAD

**Discussion by pixelcmtd** [dotnet/csharplang#2459] [Language Ideas] [answered]
# Long version explaining everything
For the last 1-2 years I've been using C#, C and friends alongside each other and I think I found the most annoying thing about C#, it's CS0031. Or at least CS0031 is a symptom of a problem that exists in C++, too. It is the problem of not being able to write **simple** code. In C I can do something like
```c
FILE *f = open_file();
if(f) read_from(f);
else return error;
```
while in C# I would have to do something like
```cs
try
{
    FileStream f = OpenFile();
    ReadFrom(f);
}
catch
{
    return error;
}
```
And that isn't even the worst part about it, because you could still argue that exceptions are better than returning zero values. But there are many use-cases where you could produce just simpler and functionally 100% equal code, I just grep-ed through some projects of mine and the "== 0" syndrome occures a pretty often, just all the `== 0`, `!= 0` and `== 1` is only there because of CS0031. `variable == 0` can just be `!variable`, `variable != 0` can be `variable` and `variable == 1` is usually just another way for saying `variable != 0`, which is `variable`. (for classes or nullables replace all '0's by "null"s)  
My conclusion from all of this is that for reducing the anger of C# developers it should be allowed to write `while(1)` and `if(!f) return error;`, at least with some kind of preprocessor instruction or compiler flag.

# TLDR
It has caused many of my angry moments that in C# you can't use your non-binary values (actually all non-`bool` values) as booleans, while C allows you to check if something is 0/null by using `if(variable)`, you **must** use `if(variable == null)` or `if(variable == 0)` in C#. I think there should be a preprocessor instruction or compiler flag for that.

**Comment by CyrusNajmabadi** [comment, marked as answer]
This was a design decision from the beginning due to how often this ends up being a bug for people.  I myself just had such a bug because i wrote `if (existing = newVal)` in C++ because i was intending to do an `==` check, but ended up with an assignment, with a conditional check on the value of that assignment.

> I think there should be a preprocessor instruction or compiler flag for that.

This would create 'dialects' of the language, and that's something the language team does not want to do (esp. for something as narrow as this).

**Comment by HaloFour** [comment]
You can emulate this behavior by returning a `Result<T>`-like type that defines an `operator true`.

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABADAAmwEYBuAWACg9CiA6ASQHlyLLsBmfWBAVwGMY+AEpwoPADYwAPABUAfPgDelfKsKdgAe03j8AZX59RUJfgDmcGCXwBfSgEgO+GfgBqEcTzimLV2yrUnAFEADyMABxgAS00AO3wAMQgozwRvRXNLazsKNXwA1XCEKIA3CHhhUQkYAAoXMs84AEolArz9Q2N8AF58GF44Fna1d0ae/AavIeHE5NTvXtiJcWnVHLy2otLy7xExSRrQiOi4/DgwuEiY2JblXJmDPiMoE16AMw8oQba80a9xlBwD7VVbtJIpHhpcbnY7XUHrNRtJzEABs+C0OnwmnCiHKmgQfQGNT21VkCjS+xgLW65Kqkloj2eUFByKIaIxumxuJg+PwH3EX2JdOk8nwFOq1IUAEJxfTGcZQUjOKj8FEwOFxFE+FFBFyEHiCSTJGS6hMPF5JfhYnAAO6VSkmybNRX3VSstHnDVanVYnH6nkEmRCh2i2VUnpyH5qMMMzovfAAfjFwtof28ID6AAsEJo7THwfMWDk2MqAEz4ADCrVd6hoaKN0gSKTgen6cAgYAUjBxsSb4jgNWIBCdlsBwMkLOVbMIABZKhAUAkc2Aan2W22O1wN2Bbv4a0rZ/gALI1W5R1RlAlvZvjbtwXvNmoAIje2ifTVBeSib3wNWv/ZaEQFyXTQV1XZtWzSDsmn/Z1zzOAVvBgbNcz5ZtaALSFvhrHIbCAA

**Comment by yaakov-h** [comment]
You can still write simple code if you put your mind to it... without requiring implicit conversions to boolean all over the place.

```csharp
if (TryOpenFile(out var f)) ReadFrom(f);
else return error;
```

**Comment by The-Futurist** [comment]
@chrissxYT - Chris, this is one of C#'s strengths, it is _very_ formal in areas where languages like C and C++ are too informal. You seem to be unhappy with the exception pattern too, but this is superior to error codes for a host of reasons one of them being that it's harder for a caller to disregard an error.

Your request amounts to making C# weaker in order to make it more attractive for your current coding practices.

**Comment by DavidArno** [comment]
> You can emulate this behavior by returning a `Result<T>`-like type that defines an `operator true`.

In fact, there's no need to even write such a type as libraries exist, [such as my own SuccincT library](https://github.com/DavidArno/SuccincT/wiki/Success_T_), that provide such a type for you.
